lib/clerk/rack_middleware.rb

```diff
@@ -8,48 +8,75 @@ def initialize(app)
 
     def call(env)
       req = Rack::Request.new(env)
-      token = req.cookies["__session"]
+      env["clerk"] = Proxy.new(env)
+      @app.call(env)
+    end
+  end
 
-      if token
-        sess_id = req.params["_clerk_session_id"]
-        begin
-          env["clerk_session"] = fetch_session(token, sess_id)
-        rescue Errors::Authentication => e
-          env["clerk_error"] = e
-        end
-      end
-      if sess = env["clerk_session"]
-        env["clerk_user"] = fetch_user(sess["user_id"])
+  class Proxy
+    attr_reader :session_id, :error
+    def initialize(env)
+      req = Rack::Request.new(env)
+      @token = req.cookies["__session"]
+      @session_id = req.params["_clerk_session_id"]
+      @session = nil
+      @user_id = nil
+      @user = nil
+    end
+
+    def session
+      return nil if @token.nil?
+      return @session if @session
+
+      begin
+        @session = fetch_session
+      rescue Errors::Authentication => e
+        @error = e
       end
-      @app.call(env)
+      @session
+    end
+
+    def user_id
+      @user_id ||= session.dig("user_id")
+    end
+
+    def user
+      return nil if session.nil?
+      @user ||= fetch_user(user_id)
+    end
+
+    def debug
+      (instance_variables - [:@sdk]).map do |ivar|
+        [ivar.to_s, instance_variable_get(ivar)]
+      end.to_h
     end
 
     private
-    def clerk_sdk
-      SDK.new
+    def sdk
+      @sdk ||= SDK.new
+    end
+
+    def cache_key
+      @cache_key ||= @token.split(".")[1]
     end
 
-    def fetch_session(token, sess_id)
-      cache_key = token.split(".")[1]
-      if sess_id
-        session = cached_fetch("clerk_session:#{sess_id}:#{cache_key}") do
-          sdk = clerk_sdk
-          sdk.sessions.verify_token(sess_id, token)
+    def fetch_session
+      if session_id
+        cached_fetch("clerk_session:#{session_id}:#{cache_key}") do
+          sdk.sessions.verify_token(session_id, @token)
         end
       else
-        session = cached_fetch("clerk_session:#{cache_key}") do
-          sdk = clerk_sdk
-          client = sdk.clients.verify_token(token)
-          sess_id = client["last_active_session_id"]
+        cached_fetch("clerk_session:#{cache_key}") do
```

Comment:
@teoulas Is caching necessary in the middleware? Shouldn't we always verify with the server at this point?

Reply:
Not necessary, but if you hit the API on every request it adds up. It's ~150ms per API call, which means +150ms for the session and +150ms for the user. We can take it out, but I think it's useful for now until we convert to networkless.

Reply:
I am afraid that we might introduce bugs by caching the verification response especially if session has changes in another tab, or via the Backend API. Can we keep it but make it an opt in feature via configuration?

Reply:
Sounds good. It is very easy to change the default. It's only one line: https://github.com/clerkinc/clerk-sdk-ruby/blob/main/lib/clerk/railtie.rb#L13

Reply:
I see, although that's a rails wide setting. Adding a control to allow that give us maximum configurability.

Reply:
It is configured to `Rails.cache` in the Railtie. Otherwise you can set it to whatever you like (by default it's not set, so no cache). See: https://github.com/clerkinc/clerk-sdk-ruby/blob/main/lib/clerk/rack_middleware.rb#L58-L64

Reply:
Gone https://github.com/clerkinc/clerk-sdk-ruby/pull/3/commits/c0d764fcd4ae45adb6628abcb1c5eebf21b47b02